IMOD/manpages/montagesize.man
.na
.nh
.TH montagesize 1 4.6.34 IMOD
.SH NAME
montagesize - Determine dimensions of a montaged image file
.SH SYNOPSIS
montagesize  input_image_file [input_piece_list_file]
.SH DESCRIPTION
Montagesize will determine the X, Y, and Z dimensions of a montaged
image file, from piece coordinates that are contained either in the
the file header, in a separate piece list file, or in an associated
metadata file (i.e., a file with the same name as the image file plus
".mdoc", as produced by SerialEM).
.P
The file names are specified exclusively as command line arguments:
first the image file name, then the piece list file name, if any.
If there is one argument, the program attempts to read the
coordinates from the image file header, or from an associated metadata
file.
.P
.SH HISTORY
.nf
Written by David Mastronarde, December 1999
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/Etomo/src/etomo/util/Montagesize.java
package etomo.util;

import java.io.File;
import java.io.IOException;
import java.util.Hashtable;

import etomo.ApplicationManager;
import etomo.BaseManager;
import etomo.process.ProcessMessages;
import etomo.process.SystemProgram;
import etomo.type.AxisID;
import etomo.type.ConstEtomoNumber;
import etomo.type.EtomoNumber;
import etomo.type.FileType;

/**
 * <p>Description: Runs montagesize on .st files.  Creates once instance per file
 * Rereads only when file has changed.</p>
 * 
* <p>Copyright: Copyright 2005 - 2015 by the Regents of the University of Colorado</p>
* <p/>
* <p>Organization: Dept. of MCD Biology, University of Colorado</p>
*
* @version $Id$
 */
public class Montagesize {
  private static final String EXT = ".pl";
  //
  // n'ton member variables
  //
  private static Hashtable instances = new Hashtable();
  //
  // member variables to prevent unnecessary reads
  //
  private FileModifiedFlag modifiedFlag;
  //
  // other member variables
  //
  private EtomoNumber x = new EtomoNumber(EtomoNumber.Type.INTEGER);
  private EtomoNumber y = new EtomoNumber(EtomoNumber.Type.INTEGER);
  private EtomoNumber z = new EtomoNumber(EtomoNumber.Type.INTEGER);

  private final File file;
  private final String propertyUserDir;
  private final AxisID axisID;

  private boolean fileExists = false;
  String[] commandArray = null;
  private boolean ignorePieceListFile = false;
  private int exitValue = -1;

  //
  // n'ton functions
  //
  /**
   * private constructor
   */
  private Montagesize(String propertyUserDir, File file, AxisID axisID) {
    this.propertyUserDir = propertyUserDir;
    this.file = file;
    this.axisID = axisID;
    modifiedFlag = new FileModifiedFlag(file);
  }

  /**
   * Function to get an instance of the class
   * @param directory
   * @param datasetName
   * @param axisID
   * @return
   */
  public static Montagesize getInstance(BaseManager manager, AxisID axisID,
    final FileType fileType) {
    File keyFile = fileType.getFile(manager, axisID);
    String key = makeKey(keyFile);
    Montagesize montagesize = (Montagesize) instances.get(key);
    if (montagesize == null) {
      return createInstance(manager.getPropertyUserDir(), key, keyFile, axisID);
    }
    montagesize.setToDefaults();
    return montagesize;
  }

  /**
   * Function to get an instance of the class
   * @param directory
   * @param datasetName
   * @param axisID
   * @return
   */
  public static Montagesize getInstance(String fileLocation, String filename,
    AxisID axisID) {
    File keyFile = Utilities.getFile(fileLocation, filename);
    String key = makeKey(keyFile);
    Montagesize montagesize = (Montagesize) instances.get(key);
    if (montagesize == null) {
      return createInstance(fileLocation, key, keyFile, axisID);
    }
    montagesize.setToDefaults();
    return montagesize;
  }

  /**
   * Function to create and save an instance of the class.  Just returns the
   * instance if it already exists.
   * @param key
   * @param file
   * @param axisID
   * @return
   */
  private static synchronized Montagesize createInstance(String propertyUserDir,
    String key, File file, AxisID axisID) {
    Montagesize montagesize = (Montagesize) instances.get(key);
    if (montagesize != null) {
      return montagesize;
    }
    montagesize = new Montagesize(propertyUserDir, file, axisID);
    instances.put(key, montagesize);
    montagesize.selfTestInvariants();
    return montagesize;
  }

  /**
   * Make a unique key from a file
   * @param file
   * @return
   */
  private static String makeKey(File file) {
    return file.getAbsolutePath();
  }

  public String getFile() {
    return file.getAbsolutePath();
  }

  //
  // other functions
  //
  /**
   * construct a piece list file
   * @return
   */
  private File makePieceListFile() {
    String filePath = file.getAbsolutePath();
    int extensionIndex = filePath.lastIndexOf(".");
    if (extensionIndex == -1 || extensionIndex == 0) {
      return new File(filePath + EXT);
    }
    return new File(filePath.substring(0, extensionIndex) + EXT);
  }

  public boolean pieceListFileExists() {
    return makePieceListFile().exists();
  }

  /**
   * reset results
   *
   */
  private void reset() {
    fileExists = false;
    x.reset();
    y.reset();
    z.reset();
    commandArray = null;
    exitValue = -1;
  }

  private final void buildCommand() {
    if (commandArray != null) {
      return;
    }
    if (file == null) {
      return;
    }
    File pieceListFile = makePieceListFile();
    if (pieceListFile.exists() && !ignorePieceListFile) {
      commandArray = new String[3];
    }
    else {
      commandArray = new String[2];
    }
    commandArray[0] = ApplicationManager.getIMODBinPath() + "montagesize";
    commandArray[1] = file.getAbsolutePath();
    if (pieceListFile.exists() && !ignorePieceListFile) {
      // bug# 1336
      commandArray[2] = pieceListFile.getAbsolutePath();
    }
  }

  private void setToDefaults() {
    setIgnorePieceListFile(false);
  }

  public synchronized void setIgnorePieceListFile(final boolean input) {
    if (ignorePieceListFile != input) {
      modifiedFlag.reset();
      reset();
    }
    ignorePieceListFile = input;
  }

  public int getExitValue() {
    return exitValue;
  }

  /**
   * run montagesize on the file
   * @throws IOException
   * @throws InvalidParameterException
   * @returns true if attempted to read
   */
  public synchronized boolean read(BaseManager manager)
    throws IOException, InvalidParameterException {
    if (file.isDirectory()) {
      throw new IOException(file + "is not a file.");
    }
    if (!file.exists()) {
      reset();
      return false;
    }
    fileExists = true;
    // If the file hasn't been modified, don't reread
    if (!modifiedFlag.isModifiedSinceLastRead()) {
      return false;
    }
    // put first timestamp after decide to read
    Utilities.timestamp("read", "montagesize", file, Utilities.STARTED_STATUS);
    // Run the montagesize command on the file.
    buildCommand();
    SystemProgram montagesize =
      new SystemProgram(manager, propertyUserDir, commandArray, axisID);
    modifiedFlag.setReadingNow();
    montagesize.run();

    exitValue = montagesize.getExitValue();
    if (exitValue != 0) {
      String[] stdOutput = montagesize.getStdOutput();
      if (stdOutput != null && stdOutput.length > 0) {
        ProcessMessages messages = montagesize.getProcessMessages();
        if (messages.size(ProcessMessages.MessageType.ERROR) > 0) {
          String message = "montagesize returned an error while reading"
            + file.getAbsolutePath() + ":\n";
          for (int i = 0; i < messages.size(ProcessMessages.MessageType.ERROR); i++) {
            message = message + messages.get(ProcessMessages.MessageType.ERROR, i) + "\n";
          }
          Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
          throw new InvalidParameterException(message);
        }
      }
    }
    // Throw an exception if the file can not be read
    String[] stdError = montagesize.getStdError();
    if (stdError != null && stdError.length > 0) {
      String message =
        "montagesize returned an error while reading" + file.getAbsolutePath() + ":\n";
      for (int i = 0; i < stdError.length; i++) {
        message = message + stdError[i] + "\n";
      }
      Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
      throw new InvalidParameterException(message);
    }

    // Parse the output
    String[] stdOutput = montagesize.getStdOutput();
    if (stdOutput == null || stdOutput.length < 1) {
      Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
      throw new IOException(
        "montagesize returned no data while reading" + file.getAbsolutePath());
    }

    for (int i = 0; i < stdOutput.length; i++) {
      // Parse the size of the data
      // Note the initial space in the string below
      String outputLine = stdOutput[i].trim();
      if (outputLine.startsWith("Total NX, NY, NZ:")) {
        String[] tokens = outputLine.split("\\s+");
        if (tokens.length < 7) {
          Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
          throw new IOException(
            "Montagesize returned less than three parameters for image size while reading"
              + file.getAbsolutePath());
        }
        x.set(tokens[4]);
        y.set(tokens[5]);
        z.set(tokens[6]);
        if (!x.isValid() || x.isNull()) {
          Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
          throw new NumberFormatException(
            "NX is not set, token is " + tokens[4] + "\n" + x.getInvalidReason());
        }
        if (!y.isValid() || y.isNull()) {
          Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
          throw new NumberFormatException(
            "NY is not set, token is " + tokens[5] + "\n" + y.getInvalidReason());
        }
        if (!z.isValid() || z.isNull()) {
          Utilities.timestamp("read", "montagesize", file, Utilities.FAILED_STATUS);
          throw new NumberFormatException(
            "NZ is not set, token is " + tokens[6] + "\n" + z.getInvalidReason());
        }
      }
    }
    Utilities.timestamp("read", "montagesize", file, Utilities.FINISHED_STATUS);
    return true;
  }

  /**
   * 
   * @return
   */
  public ConstEtomoNumber getX() {
    return x;
  }

  /**
   * 
   * @return
   */
  public ConstEtomoNumber getY() {
    return y;
  }

  /**
   * 
   * @return
   */
  public ConstEtomoNumber getZ() {
    return z;
  }

  /**
   * 
   * @return
   */
  public boolean isFileExists() {
    return fileExists;
  }

  //
  // self test functions
  //
  void selfTestInvariants() {
    if (!Utilities.isSelfTest()) {
      return;
    }
    if (instances == null) {
      throw new IllegalStateException("instances is null");
    }
    if (file == null || file.isDirectory()) {
      throw new IllegalStateException("file is null");
    }
    String key = makeKey(file);
    if (key == null || key.matches("\\s*")) {
      throw new IllegalStateException(
        "unable to make key: filename=" + file.getAbsolutePath());
    }
    Montagesize montagesize = (Montagesize) instances.get(key);
    if (montagesize == null) {
      throw new IllegalStateException("this instance is not in instances: key=" + key);
    }
  }

  public String toString() {
    return getClass().getName() + "[" + super.toString() + paramString() + "]";
  }

  String paramString() {
    return ",file=" + file + ",fileExists=" + fileExists + ",x=" + x + ",y=" + y + ",z="
      + z + ",axisID=" + axisID;
  }
}
/**
 * <p> $Log$
 * <p> Revision 1.21  2010/03/27 05:13:39  sueh
 * <p> bug# 1336 Fixed the command line.
 * <p>
 * <p> Revision 1.20  2010/02/17 05:05:58  sueh
 * <p> bug# 1301 Using manager instead of manager key for popping up
 * <p> messages.
 * <p>
 * <p> Revision 1.19  2009/03/17 00:46:43  sueh
 * <p> bug# 1186 Pass managerKey to everything that pops up a dialog.
 * <p>
 * <p> Revision 1.18  2009/02/13 02:39:32  sueh
 * <p> bug# 1152 Added a getInstance function which can construct a file from
 * <p> the file location and the file name.
 * <p>
 * <p> Revision 1.17  2008/11/21 22:39:31  sueh
 * <p> bug# 1138 The command line is too long so do not use the absolute path
 * <p> for the .pl file.
 * <p>
 * <p> Revision 1.16  2007/12/26 22:41:29  sueh
 * <p> bug# 1052 Moved argument handling from EtomoDirector to a separate class.
 * <p>
 * <p> Revision 1.15  2007/09/07 00:30:43  sueh
 * <p> bug# 989 Using a public INSTANCE to refer to the EtomoDirector singleton
 * <p> instead of getInstance and createInstance.
 * <p>
 * <p> Revision 1.14  2007/08/16 16:38:23  sueh
 * <p> bug# 1035 Added getFile().
 * <p>
 * <p> Revision 1.13  2007/02/05 23:47:15  sueh
 * <p> bug# 962 Moved EtomoNumber type info to inner class.
 * <p>
 * <p> Revision 1.12  2005/11/02 22:15:51  sueh
 * <p> bug# 754 Parsing errors and warnings inside ProcessMessages.
 * <p>
 * <p> Revision 1.11  2005/10/28 18:58:09  sueh
 * <p> bug# 747 Standardizing SystemProgram message parsing.
 * <p>
 * <p> Revision 1.10  2005/10/27 00:38:32  sueh
 * <p> bug# 725 Added buildCommand().  Changed String filename to File file.
 * <p>
 * <p> Revision 1.9  2005/09/09 21:48:24  sueh
 * <p> bug# 532 Handling null from stderr and stdout.
 * <p>
 * <p> Revision 1.8  2005/08/27 22:44:01  sueh
 * <p> bug# 532 In Utilities.timestamp() change the int status to String status,
 * <p> since it doesn't have to be compared.
 * <p>
 * <p> Revision 1.7  2005/07/29 00:55:23  sueh
 * <p> bug# 709 Going to EtomoDirector to get the current manager is unreliable
 * <p> because the current manager changes when the user changes the tab.
 * <p> Passing the manager where its needed.
 * <p>
 * <p> Revision 1.6  2005/06/21 00:54:03  sueh
 * <p> bug# 522 Changed File file to String filename.  Avoid holding on to the file
 * <p> in case that would make problems in Windows.  Read() returns true if it
 * <p> attempts to read.  Overiding toString().
 * <p>
 * <p> Revision 1.5  2005/06/20 17:05:36  sueh
 * <p> bug# 522 Changed  Montagesize so that read() is not called
 * <p> automatically.
 * <p>
 * <p> Revision 1.4  2005/06/17 20:04:38  sueh
 * <p> bug# 685 Added timestamps to read().
 * <p>
 * <p> Revision 1.3  2005/04/25 21:42:49  sueh
 * <p> bug# 615 Passing the axis where a command originates to the message
 * <p> functions so that the message will be popped up in the correct window.
 * <p> This requires adding AxisID to many objects.
 * <p>
 * <p> Revision 1.2  2005/03/29 19:57:25  sueh
 * <p> bug# 623 Added fileExists state.  Added pieceListFile to the command
 * <p> string, if it exists.
 * <p>
 * <p> Revision 1.1  2005/03/08 02:01:47  sueh
 * <p> bug# 533 Object to call montagesize.
 * <p> </p>
 */
